### ISSUE TITLE
rnn_cell.py improvements

### ISSUE THREAD

**Issue by alrojo** [tensorflow/tensorflow#2150]
## Motivation
Current implementation of [rnn_cell.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py) does not support custom initialization on a gate and input-to-hidden/hidden-to-hidden level (like setting forgetgate bias to 0 while leaving updategate bias at 1, etc.). Further when debugging the gates in TensorBoard, the matrix is represented as one large matrix, which makes it difficult to see whats happening inside a specific matrix (e.g. hidden-to-hidden) of a specific gate (e.g. forgetgate).

In recurrent neural networks (RNNs) GRU and LSTM uses various gates with separate weights, for both hidden-to-hidden and input-to-hidden, to computing steps in a recurrent sequence.
To optimize computational speed these gates, and their separate weights, are often stacked and computed simultaneous at every step.
In TensorFlows rnn_cell.py the [GRUCell](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L123) and [BasicLSTMCell](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L153) are implemented using [linear](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L677) to handle weigths and computation hereof.
However, the implementation of _linear_ does not initialize separate matrices for each gate, but initializes the gates, and their input-to-hidden/hidden-to-hidden matrices, as one big matrix for [weights](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L712) and [bias](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py#L719).

## Proposal
Implementing separate initialization of gates, and their input-to-hidden/hidden-to-hidden matrices, and concatenating these gates. This allows custom initializatio, TensorBoard information on hid_in/hid_hid/bias for every gate and still retains the advantage of weights in a large matrix.

## Implementation
With minimal rewriting of the current structure in [rnn_cell.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell.py) I implemented a [lasagne-gate](https://lasagne.readthedocs.org/en/latest/modules/layers/recurrent.html#lasagne.layers.Gate) like structure, made a new _linear_ function and made some minor changes to _GRUCell_.
All of these changes should, with minor modifications, work for _BasicLSTM_ as well.

Do notice that code below is for my own purpose, it is not rigorously tested yet.

First, the gate to hold initialization for every weight matrix/bias in a gate (notice it also handles LSTM's by having W_cell)
```python
class Gate(object):
  """Gate to handle to handle initialization"""  

  def __init__(self, W_in=init_ops.random_normal_initializer(stddev=0.1),
               W_hid=init_ops.random_normal_initializer(stddev=0.1),
               W_cell=init_ops.random_normal_initializer(stddev=0.1),
               b=init_ops.constant_initializer(0.),
               activation=None):
    self.W_in = W_in
    self.W_hid = W_hid
    # Don't store a cell weight vector when cell is None
    if W_cell is not None:
        self.W_cell = W_cell
    if b is not None:
      self.b = b
    # For the activation, if None is supplied, use identity
    if activation is None:
        self.activation = control_flow_ops.identity
    else:
        self.activation = activation
```

A modified GRU cell to handle weigths (took only minimal modification to __init__ and __call__)
```python
class GRUCell(rnn_cell.RNNCell):
  """Gated Recurrent Unit cell (cf. http://arxiv.org/abs/1406.1078)."""

  def __init__(self, num_units, input_size=None,
               resetgate=Gate(W_cell=None, activation=sigmoid),
               updategate=Gate(W_cell=None, activation=sigmoid),
               candidategate=Gate(W_cell=None, activation=tanh)):
    self._num_units = num_units
    self._input_size = num_units if input_size is None else input_size
    self._resetgate = resetgate
    self._updategate = updategate
    self._candidategate = candidategate

  @property
  def input_size(self):
    return self._input_size

  @property
  def output_size(self):
    return self._num_units

  @property
  def state_size(self):
    return self._num_units

  def __call__(self, inputs, state, scope=None):
    """Gated recurrent unit (GRU) with nunits cells."""
    with vs.variable_scope(scope or type(self).__name__):  # "GRUCell"
      with vs.variable_scope("Gates"):  # Reset gate and update gate.
        # We start with bias of 1.0 to not reset and not update.
        r, u = array_ops.split(1, 2, Modified_linear([inputs, state],
          [(self._num_units, "Reset", self._resetgate),
           (self._num_units, "Update", self._updategate)]))
        r, u = self._resetgate.activation(r), self._updategate.activation(u)
      with vs.variable_scope("Candidate"):
        c = Modified_linear([inputs, r * state],
          (self._num_units, "Candidate", self._candidategate))
        c = self._candidategate.activation(c)
      new_h = u * state + (1 - u) * c
    return new_h, new_h
```

I found _linear_ required the largest amount of rewriting, however I have tried to keep the original structure and functionality intact.

```python
def Modified_linear(args, output, scope=None):
  """Modified linear takes args and output.
     Args is same as in linear, but output is a tuple consisting of:
     output_size, name of gate, gate object (with all initializations)
  """
  if args is None or (isinstance(args, (list, tuple)) and not args):
    raise ValueError("`args` must be specified")
  if not isinstance(args, (list, tuple)):
    args = [args]
  if not isinstance(output, list):
    output = [output]
  shapes = [a.get_shape().as_list() for a in args]
  for shape in shapes:
    if len(shape) != 2:
      raise ValueError("Linear is expecting 2D arguments: %s" % str(shapes))
    if not shape[1]:
      raise ValueError("Linear expects shape[1] of arguments: %s" % str(shapes))

  matrices = []
  biases = []
  with vs.variable_scope(scope or "Linear"):
    for output_size, name, gate in output: # loops over every gate
      with vs.variable_scope(name):
        W_in = vs.get_variable("W_in", [args[0].get_shape()[1], output_size],
          initializer=gate.W_in)
        W_hid = vs.get_variable("W_hid", [args[1].get_shape()[1], output_size],
          initializer=gate.W_hid)
        if hasattr(gate, 'b'):
          b = vs.get_variable("Bias", [output_size],
            initializer=gate.b)
          biases.append(b)
        if hasattr(gate, "W_cell"):
          pass
          # do some LSTM stuff ...
        else:
          matrix = array_ops.concat(0, [W_in, W_hid]) # concats all matrices
        matrices.append(matrix)

  total_matrix = array_ops.concat(1, matrices) # concats across gates
  res = math_ops.matmul(array_ops.concat(1, args), total_matrix) # computes the results

  if biases is not []:
    total_bias = array_ops.concat(0, biases) # concats across gates biases
    if total_matrix.get_shape()[1] != total_bias.get_shape()[0]:
      raise ValueError('Must have same output dimensions for W and b')
    res += total_bias
  return res
```

## Questions

- Would this be of interest for a PR to _rnn_cell.py_? (given further development of code and _BasicLSTMCell_ implementation)
- General comments/thoughts would be much appreciated

**Comment by ebrevdo** [comment]
Two comments:

1. Gate seems like an overkill class.
If you want to allow setting the nonlinearity and initialization of a particular gate, why not just accept and store these parameters directly?

2. We're deprecating linear() in favor of the layers functions, like fully_connected. In fact, linear can and will disappear as soon as can make this happen.

**Comment by alrojo** [comment]
@ebrevdo I agree with the gates, also they would only work for basic GRU/LSTM.

I have updated the code based on your comments (though omitted fail-safes for clarifying functionality).

Would this be compatible with your plans on `GRUCell` and `BasicLSTMCell`?
```python
class GRUCell(rnn_cell.RNNCell):
  """Gated Recurrent Unit cell (cf. http://arxiv.org/abs/1406.1078)."""

  def __init__(self, num_units, input_size=None,
               reset_W_in = init_ops.random_normal_initializer(stddev=0.1),
               reset_W_hid = init_ops.random_normal_initializer(stddev=0.1),
               reset_b = init_ops.constant_initializer(0.),
               reset_activation = sigmoid, 
               update_W_in = init_ops.random_normal_initializer(stddev=0.1),
               update_W_hid = init_ops.random_normal_initializer(stddev=0.1),
               update_b = init_ops.constant_initializer(0.),
               update_activation = sigmoid,
               candidate_W_in = init_ops.random_normal_initializer(stddev=0.1),
               candidate_W_hid = init_ops.random_normal_initializer(stddev=0.1),
               candidate_b = init_ops.constant_initializer(0.),
               candidate_activation = tanh):
    self._num_units = num_units
    self._input_size = num_units if input_size is None else input_size
    self._reset_W_in = reset_W_in
    self._reset_W_hid = reset_W_hid
    self._reset_b = reset_b
    self._reset_activation = reset_activation
    self._update_W_in = update_W_in
    self._update_W_hid = update_W_hid
    self._update_b = update_b
    self._update_activation = update_activation
    self._candidate_W_in = candidate_W_in
    self._candidate_W_hid = candidate_W_hid
    self._candidate_b = candidate_b
    self._candidate_activation = candidate_activation

  @property
  def input_size(self):
    return self._input_size

  @property
  def output_size(self):
    return self._num_units

  @property
  def state_size(self):
    return self._num_units

  def _compute(self, args, gate):
    name, W_in_init, W_hid_init, b_init = gate
    with vs.variable_scope(name):
      W_in = vs.get_variable("W_in",
        [args[0].get_shape()[1], self._num_units],
        initializer=W_in_init)
      W_hid = vs.get_variable("W_hid",
        [args[1].get_shape()[1], self._num_units],
        initializer=W_hid_init)
      b = vs.get_variable("Bias", [self._num_units],
        initializer=b_init)
      matrix = array_ops.concat(0, [W_in, W_hid])
    return matrix, b

  def __call__(self, inputs, state, scope=None):
    """Gated recurrent unit (GRU) with nunits cells."""
    args = [inputs, state]
    with vs.variable_scope(scope or type(self).__name__):  # "GRUCell"
      with vs.variable_scope("Gates"):  # Reset gate and update gate.
        matrices = []
        biases = []
        gates = [
          ("Reset", self._reset_W_in, self._reset_W_hid, self._reset_b),
          ("Update", self._update_W_in, self._update_W_hid, self._update_b)]
        for gate in gates:
          matrix, bias = self._compute(args, gate)
          matrices.append(matrix)
          biases.append(bias)
        total_matrix = array_ops.concat(1, matrices)
        total_bias = array_ops.concat(0, biases)
        res_gates = math_ops.matmul(array_ops.concat(1, args), total_matrix)
        res_gates += total_bias
        r, u = array_ops.split(1, 2, res_gates)
        r, u = self._reset_activation(r), self._update_activation(u)
      with vs.variable_scope("Candidate"):
        candidate = ("Candidate", self._candidate_W_in, self._candidate_W_hid,
            self._candidate_b)
        matrix, bias = self._compute([inputs, r * state], candidate)
        c = math_ops.matmul(array_ops.concat(1, args), matrix) + bias
        c = self._candidate_activation(c)
      new_h = u * state + (1 - u) * c
    return new_h, new_h
```

**Comment by alrojo** [comment]
Any news @ebrevdo?

**Comment by ebrevdo** [comment]
We currently use single fused weight and bias variables in order to speed up computation.  Having separate initializers would lead to a significant slowdown.  You are welcome to write your own RNNCell which creates separate weight variables, each with their own initializers.

**Comment by ebrevdo** [comment]
Closing this for now; as it sounds straightforward for you to create this custom LSTMCell in your own repository.

**Comment by xyzhou-puck** [comment]
So does anyone test the implementation above? It would be very useful if we could easily implement our own RNNCells. :)
